## Sourcing open data

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
path = r'/Users/stadtsalat/Documents/Boat Sales'

In [3]:
# Importing boat_data.csv
boat_data = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'boat_data.csv'))

In [4]:
# Checking the boat_data df
boat_data.head()

Price            Boat Type          Manufacturer                 Type  \
0   CHF 3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   EUR 3490  Center console boat     Terhi power boats  new boat from stock   
2   CHF 3770           Sport Boat    Marine power boats  new boat from stock   
3  DKK 25900           Sport Boat    Pioner power boats  new boat from stock   
4   EUR 3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90            NaN   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00            NaN   
4        2019    3.55   1.46      Aluminium   

                                     Location  Number of views last 7 days  
0      Switzerland » Lake Geneva » Vésenaz                          226  
1                    Germany » Bönningstedt                           75  
2  Switzerland » Lake of Zurich » Stäfa ZH                          124  
3                        Denmark » Svendborg                           64  
4               Germany » Bayern » München                           58

In [5]:
# Checking the info of the df
boat_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9888 entries, 0 to 9887
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Price                        9888 non-null   object 
 1   Boat Type                    9888 non-null   object 
 2   Manufacturer                 8550 non-null   object 
 3   Type                         9882 non-null   object 
 4   Year Built                   9888 non-null   int64  
 5   Length                       9879 non-null   float64
 6   Width                        9832 non-null   float64
 7   Material                     8139 non-null   object 
 8   Location                     9852 non-null   object 
 9   Number of views last 7 days  9888 non-null   int64  
dtypes: float64(2), int64(2), object(6)
memory usage: 772.6+ KB


In [6]:
# Checking the size of the df
boat_data.shape

(9888, 10)

### Data quality and consistency checks

In [7]:
# Checking for duplicates
df_dups = boat_data[boat_data.duplicated()]

In [8]:
df_dups

Empty DataFrame
Columns: [Price, Boat Type, Manufacturer, Type, Year Built, Length, Width, Material, Location, Number of views last 7 days]
Index: []

There are no duplicates

In [9]:
# Renaming column 'Number of views last 7 days'
boat_data = boat_data.rename(columns={'Number of views last 7 days': 'Views last 7 days'})


In [10]:
# Checking the column names
boat_data.head(1)

Price    Boat Type          Manufacturer                 Type  \
0  CHF 3337  Motor Yacht  Rigiflex power boats  new boat from stock   

   Year Built  Length  Width Material                                Location  \
0        2017     4.0    1.9      NaN  Switzerland » Lake Geneva » Vésenaz   

   Views last 7 days  
0                226

In [11]:
# Checking for missing values
boat_data.isnull().sum()

Price                   0
Boat Type               0
Manufacturer         1338
Type                    6
Year Built              0
Length                  9
Width                  56
Material             1749
Location               36
Views last 7 days       0
dtype: int64

In [12]:
# Checking the percentage of missing values for every column
(boat_data.isnull().sum()/len(boat_data) * 100).sort_values()


Price                 0.000000
Boat Type             0.000000
Year Built            0.000000
Views last 7 days     0.000000
Type                  0.060680
Length                0.091019
Location              0.364078
Width                 0.566343
Manufacturer         13.531553
Material             17.688107
dtype: float64

We have a lot of missing values in columns 'Material' and 'Manufacturer'. This needs to be further investigated

In [13]:
# Checking the empty cells in Manufacturer
boat_data[boat_data['Manufacturer'].isnull()]


Price     Boat Type Manufacturer                          Type  \
6      CHF 3600     Catamaran          NaN            Used boat,Unleaded   
7     DKK 24800    Sport Boat          NaN                     Used boat   
17    DKK 30000    Sport Boat          NaN                     Used boat   
19     EUR 4000    Cabin Boat          NaN            Used boat,Unleaded   
24     EUR 3900    Cabin Boat          NaN                     Used boat   
...         ...           ...          ...                           ...   
9864   EUR 4990    Sport Boat          NaN            Used boat,Unleaded   
9866  DKK 36500    Sport Boat          NaN                     Used boat   
9873   EUR 4799  Working Boat          NaN  new boat from stock,Electric   
9875   EUR 4790       Classic          NaN                     Used boat   
9884   EUR 4516    Sport Boat          NaN           new boat from stock   

      Year Built  Length  Width   Material  \
6           1999    6.20   2.38  Aluminium   
7              0    3.00    NaN        NaN   
17          1985    4.00   1.00        NaN   
19          2011    6.37   2.31        GRP   
24          1985    6.53   2.47        NaN   
...          ...     ...    ...        ...   
9864           0    4.60   1.80        GRP   
9866        1985    4.00   1.00        NaN   
9873        2019    3.64   1.37        NaN   
9875        1962    4.10   1.50       Wood   
9884           0    4.17   1.68        GRP   

                                       Location  Views last 7 days  
6     Switzerland » Neuenburgersee » Yvonand                 474  
7                          Denmark » Svendborg                134  
17                         Denmark » Svendborg                124  
19          Germany » Bayern » Forchheim/Ofr.                330  
24                  France » Marseille, France                383  
...                                         ...                ...  
9864                                    Germany                185  
9866                       Denmark » Svendborg                 96  
9873          Germany » Bayern » Boote Jochum                 41  
9875                         Austria » Pöllau                188  
9884              Germany » Hamburg » HAMBURG                 94  

[1338 rows x 10 columns]

I cannot identify a specific correlation between the missing values and the rest of the columns. Since, the missing values are 13.5% of the total values, I decided to not impute it. I will rename the NaN values to Unknown. 

In [14]:
# Checking the empty cells in Material
boat_data[boat_data['Material'].isnull()]

Price     Boat Type          Manufacturer  \
0      CHF 3337   Motor Yacht  Rigiflex power boats   
3     DKK 25900    Sport Boat    Pioner power boats   
7     DKK 24800    Sport Boat                   NaN   
8      EUR 3333  Fishing Boat  Crescent power boats   
9      EUR 3300  Pontoon Boat     Whaly power boats   
...         ...           ...                   ...   
9867  DKK 36500  Fishing Boat    Hunter power boats   
9873   EUR 4799  Working Boat                   NaN   
9879   CHF 4990    Sport Boat    Pioner power boats   
9883   CHF 4900    Sport Boat   Sea Ray power boats   
9886   EUR 4300  Pontoon Boat     Whaly power boats   

                              Type  Year Built  Length  Width Material  \
0              new boat from stock        2017    4.00   1.90      NaN   
3              new boat from stock        2020    3.00   1.00      NaN   
7                        Used boat           0    3.00    NaN      NaN   
8              new boat from stock        2019    3.64   1.37      NaN   
9              new boat from stock        2018    4.35   1.73      NaN   
...                            ...         ...     ...    ...      ...   
9867                     Used boat        2000    5.00   1.00      NaN   
9873  new boat from stock,Electric        2019    3.64   1.37      NaN   
9879             new boat on order           0    4.11   1.73      NaN   
9883            Used boat,Unleaded        1987    6.30   2.44      NaN   
9886           new boat from stock        2018    4.37   1.89      NaN   

                                           Location  Views last 7 days  
0            Switzerland » Lake Geneva » Vésenaz                226  
3                              Denmark » Svendborg                 64  
7                              Denmark » Svendborg                134  
8     Germany » Bayern » Boote+service Oberbayern                 45  
9                             Italy » Dormelletto                 180  
...                                             ...                ...  
9867                           Denmark » Svendborg                 87  
9873              Germany » Bayern » Boote Jochum                 41  
9879                        Switzerland » Safenwil                280  
9883       Switzerland » Lago Maggiore » Riazzino               1116  
9886                          Italy » Dormelletto                 266  

[1749 rows x 10 columns]

Same here, I cannot identify a specific correlation between the missing values and the rest of the columns. Since, the missing values are 17.6% of the total values, I decided to not impute it. I will rename the NaN values to Unknown. 

The rest of the NaNs in the other columns can be dropped as they are less than 1% of the datasate

In [15]:
# Replacing NaNs with 'Unknown' in Manufacturer and Material columns
boat_data[['Manufacturer', 'Material']] = boat_data[['Manufacturer', 'Material']].fillna(value='Unknown')


In [16]:
(boat_data.isnull().sum()/len(boat_data) * 100).sort_values()


Price                0.000000
Boat Type            0.000000
Manufacturer         0.000000
Year Built           0.000000
Material             0.000000
Views last 7 days    0.000000
Type                 0.060680
Length               0.091019
Location             0.364078
Width                0.566343
dtype: float64

In [17]:
# Dropping the rest of the Nans
boat_data = boat_data.dropna()


In [18]:
# Checking the percentage of missing values for every column
(boat_data.isnull().sum()/len(boat_data) * 100).sort_values()

Price                0.0
Boat Type            0.0
Manufacturer         0.0
Type                 0.0
Year Built           0.0
Length               0.0
Width                0.0
Material             0.0
Location             0.0
Views last 7 days    0.0
dtype: float64

There are no more missing values

In [19]:
# Checking the df
boat_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9790 entries, 0 to 9887
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price              9790 non-null   object 
 1   Boat Type          9790 non-null   object 
 2   Manufacturer       9790 non-null   object 
 3   Type               9790 non-null   object 
 4   Year Built         9790 non-null   int64  
 5   Length             9790 non-null   float64
 6   Width              9790 non-null   float64
 7   Material           9790 non-null   object 
 8   Location           9790 non-null   object 
 9   Views last 7 days  9790 non-null   int64  
dtypes: float64(2), int64(2), object(6)
memory usage: 841.3+ KB


In [20]:
# Checking for mixed data types
for col in boat_data.columns.tolist():
  weird = (boat_data[[col]].applymap(type) != boat_data[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (boat_data[weird]) > 0:
    print (col)

There no mixed data types

In [21]:
c

NameError: name 'c' is not defined

In [22]:
# Checking the df
boat_data.describe()

Year Built       Length        Width  Views last 7 days
count  9790.000000  9790.000000  9790.000000        9790.000000
mean   1892.499387    11.537413     3.514004         149.366803
std     461.575903     5.970952     1.213858         152.243472
min       0.000000     1.040000     0.010000          13.000000
25%    1996.000000     7.460000     2.540000          70.000000
50%    2007.000000    10.245000     3.320000         108.000000
75%    2017.000000    13.900000     4.250000         172.000000
max    2021.000000   100.000000    25.160000        3263.000000

From the two checks above we notice a few things: 

1.The prices are in different currencies. In order to be able to easier analyse the df, I would change everything to Euros

2.The way the values in location are entered makes the analysis difficult. There, I would extract the country and the city 

3.In Year built, we have min value of '0'. We can also see the row with that value on the above Check at row 2. I cannot tell what year that boat was built, so I will just drop that row.

In [23]:
# First we need to split Price to 'Currency'and 'Price'
boat_data[['Currency','Price']]= boat_data['Price'].str.split(' ', 1, expand=True)

In [24]:
boat_data.head(5)

Price            Boat Type          Manufacturer                 Type  \
0   3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   3490  Center console boat     Terhi power boats  new boat from stock   
2   3770           Sport Boat    Marine power boats  new boat from stock   
3  25900           Sport Boat    Pioner power boats  new boat from stock   
4   3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90        Unknown   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00        Unknown   
4        2019    3.55   1.46      Aluminium   

                                     Location  Views last 7 days Currency  
0      Switzerland » Lake Geneva » Vésenaz                226      CHF  
1                    Germany » Bönningstedt                 75      EUR  
2  Switzerland » Lake of Zurich » Stäfa ZH                124      CHF  
3                        Denmark » Svendborg                 64      DKK  
4               Germany » Bayern » München                 58      EUR

In [25]:
# Checking the types of currency
boat_data['Currency'].value_counts()

EUR    8346
CHF     976
£      295
DKK     173
Name: Currency, dtype: int64

In [26]:
# Changing the Currencies to Euro with Feb 2023 exchange rates 
def price_in_euro(Price,Currency):
    if Currency=='EUR':
        return Price
    elif Currency=='CHF':
        return int(Price) * 1.01
    elif Currency=='DKK':
        return int(Price) * 0.13
    elif Currency=='£':
        return int(Price) * 1.12
boat_data['Price_in_Euro'] = boat_data.apply(lambda x: price_in_euro(x.Price, x.Currency), axis=1)

In [27]:
boat_data.head(5)

Price            Boat Type          Manufacturer                 Type  \
0   3337          Motor Yacht  Rigiflex power boats  new boat from stock   
1   3490  Center console boat     Terhi power boats  new boat from stock   
2   3770           Sport Boat    Marine power boats  new boat from stock   
3  25900           Sport Boat    Pioner power boats  new boat from stock   
4   3399         Fishing Boat    Linder power boats  new boat from stock   

   Year Built  Length  Width       Material  \
0        2017    4.00   1.90        Unknown   
1        2020    4.00   1.50  Thermoplastic   
2           0    3.69   1.42      Aluminium   
3        2020    3.00   1.00        Unknown   
4        2019    3.55   1.46      Aluminium   

                                     Location  Views last 7 days Currency  \
0      Switzerland » Lake Geneva » Vésenaz                226      CHF   
1                    Germany » Bönningstedt                 75      EUR   
2  Switzerland » Lake of Zurich » Stäfa ZH                124      CHF   
3                        Denmark » Svendborg                 64      DKK   
4               Germany » Bayern » München                 58      EUR   

  Price_in_Euro  
0       3370.37  
1          3490  
2        3807.7  
3        3367.0  
4          3399

In [28]:
# Dropping columns Price and Currency
boat_data = boat_data.drop(['Price','Currency'],axis=1)


In [29]:
boat_data.head(5)

Boat Type          Manufacturer                 Type  Year Built  \
0          Motor Yacht  Rigiflex power boats  new boat from stock        2017   
1  Center console boat     Terhi power boats  new boat from stock        2020   
2           Sport Boat    Marine power boats  new boat from stock           0   
3           Sport Boat    Pioner power boats  new boat from stock        2020   
4         Fishing Boat    Linder power boats  new boat from stock        2019   

   Length  Width       Material                                    Location  \
0    4.00   1.90        Unknown      Switzerland » Lake Geneva » Vésenaz   
1    4.00   1.50  Thermoplastic                    Germany » Bönningstedt   
2    3.69   1.42      Aluminium  Switzerland » Lake of Zurich » Stäfa ZH   
3    3.00   1.00        Unknown                        Denmark » Svendborg   
4    3.55   1.46      Aluminium               Germany » Bayern » München   

   Views last 7 days Price_in_Euro  
0                226       3370.37  
1                 75          3490  
2                124        3807.7  
3                 64        3367.0  
4                 58          3399

In [30]:
# Extracting from Location - Country and City
boat_data[['Country','City']]= boat_data['Location'].str.split('»', 1, expand=True)
boat_data[['City','Other']]= boat_data['City'].str.split('»', 1, expand=True)
boat_data['City']=boat_data['City'].str.replace('¶','',regex=True)
boat_data['City']=boat_data['City'].str.replace('Ã','',regex=True)

In [31]:
boat_data.head(5)

Boat Type          Manufacturer                 Type  Year Built  \
0          Motor Yacht  Rigiflex power boats  new boat from stock        2017   
1  Center console boat     Terhi power boats  new boat from stock        2020   
2           Sport Boat    Marine power boats  new boat from stock           0   
3           Sport Boat    Pioner power boats  new boat from stock        2020   
4         Fishing Boat    Linder power boats  new boat from stock        2019   

   Length  Width       Material                                    Location  \
0    4.00   1.90        Unknown      Switzerland » Lake Geneva » Vésenaz   
1    4.00   1.50  Thermoplastic                    Germany » Bönningstedt   
2    3.69   1.42      Aluminium  Switzerland » Lake of Zurich » Stäfa ZH   
3    3.00   1.00        Unknown                        Denmark » Svendborg   
4    3.55   1.46      Aluminium               Germany » Bayern » München   

   Views last 7 days Price_in_Euro       Country              City       Other  
0                226       3370.37  Switzerland       Lake Geneva     Vésenaz  
1                 75          3490      Germany        Bnningstedt        None  
2                124        3807.7  Switzerland    Lake of Zurich    Stäfa ZH  
3                 64        3367.0      Denmark          Svendborg        None  
4                 58          3399      Germany            Bayern     München

In [32]:
# Dropping column Other and Location
boat_data = boat_data.drop(['Other','Location'],axis=1)


In [33]:
boat_data.head(5)

Boat Type          Manufacturer                 Type  Year Built  \
0          Motor Yacht  Rigiflex power boats  new boat from stock        2017   
1  Center console boat     Terhi power boats  new boat from stock        2020   
2           Sport Boat    Marine power boats  new boat from stock           0   
3           Sport Boat    Pioner power boats  new boat from stock        2020   
4         Fishing Boat    Linder power boats  new boat from stock        2019   

   Length  Width       Material  Views last 7 days Price_in_Euro  \
0    4.00   1.90        Unknown                226       3370.37   
1    4.00   1.50  Thermoplastic                 75          3490   
2    3.69   1.42      Aluminium                124        3807.7   
3    3.00   1.00        Unknown                 64        3367.0   
4    3.55   1.46      Aluminium                 58          3399   

        Country              City  
0  Switzerland       Lake Geneva   
1      Germany        Bnningstedt  
2  Switzerland    Lake of Zurich   
3      Denmark          Svendborg  
4      Germany            Bayern

In [34]:
# Dropping the row where 'Year Built' is 0
boat_data = boat_data[boat_data['Year Built']!=0]

In [35]:
boat_data.head(5)

Boat Type          Manufacturer                 Type  Year Built  \
0          Motor Yacht  Rigiflex power boats  new boat from stock        2017   
1  Center console boat     Terhi power boats  new boat from stock        2020   
3           Sport Boat    Pioner power boats  new boat from stock        2020   
4         Fishing Boat    Linder power boats  new boat from stock        2019   
6            Catamaran               Unknown   Used boat,Unleaded        1999   

   Length  Width       Material  Views last 7 days Price_in_Euro  \
0    4.00   1.90        Unknown                226       3370.37   
1    4.00   1.50  Thermoplastic                 75          3490   
3    3.00   1.00        Unknown                 64        3367.0   
4    3.55   1.46      Aluminium                 58          3399   
6    6.20   2.38      Aluminium                474        3636.0   

        Country              City  
0  Switzerland       Lake Geneva   
1      Germany        Bnningstedt  
3      Denmark          Svendborg  
4      Germany            Bayern   
6  Switzerland    Neuenburgersee

In [36]:
# Checking the df
boat_data.describe()

Year Built       Length        Width  Views last 7 days
count  9241.000000  9241.000000  9241.000000        9241.000000
mean   2004.931176    11.717052     3.552552         150.446813
std      16.387410     5.997217     1.211919         155.106784
min    1885.000000     1.040000     0.010000          13.000000
25%    1999.000000     7.560000     2.550000          70.000000
50%    2008.000000    10.500000     3.390000         108.000000
75%    2018.000000    14.000000     4.260000         172.000000
max    2021.000000   100.000000    25.160000        3263.000000

In [37]:
# Checking the df
boat_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9241 entries, 0 to 9887
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Boat Type          9241 non-null   object 
 1   Manufacturer       9241 non-null   object 
 2   Type               9241 non-null   object 
 3   Year Built         9241 non-null   int64  
 4   Length             9241 non-null   float64
 5   Width              9241 non-null   float64
 6   Material           9241 non-null   object 
 7   Views last 7 days  9241 non-null   int64  
 8   Price_in_Euro      9241 non-null   object 
 9   Country            9241 non-null   object 
 10  City               8399 non-null   object 
dtypes: float64(2), int64(2), object(7)
memory usage: 866.3+ KB


In [38]:
# We need to change the Price_in_Euro data type to int
boat_data['Price_in_Euro'] = boat_data['Price_in_Euro'].astype('int64')

In [39]:
# Checking the df
boat_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9241 entries, 0 to 9887
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Boat Type          9241 non-null   object 
 1   Manufacturer       9241 non-null   object 
 2   Type               9241 non-null   object 
 3   Year Built         9241 non-null   int64  
 4   Length             9241 non-null   float64
 5   Width              9241 non-null   float64
 6   Material           9241 non-null   object 
 7   Views last 7 days  9241 non-null   int64  
 8   Price_in_Euro      9241 non-null   int64  
 9   Country            9241 non-null   object 
 10  City               8399 non-null   object 
dtypes: float64(2), int64(3), object(6)
memory usage: 866.3+ KB


In [40]:
# Checking for missing values
(boat_data.isnull().sum()/len(boat_data) * 100).sort_values()


Boat Type            0.000000
Manufacturer         0.000000
Type                 0.000000
Year Built           0.000000
Length               0.000000
Width                0.000000
Material             0.000000
Views last 7 days    0.000000
Price_in_Euro        0.000000
Country              0.000000
City                 9.111568
dtype: float64

In the new column 'City' we are missing 9% of the values. In this case I will replace it again with Unknown

In [41]:
# Replacing NaNs with 'Unknown' in City
boat_data[['City']] = boat_data[['City']].fillna(value='Unknown')


In [42]:
# Checking for missing values
(boat_data.isnull().sum()/len(boat_data) * 100).sort_values()


Boat Type            0.0
Manufacturer         0.0
Type                 0.0
Year Built           0.0
Length               0.0
Width                0.0
Material             0.0
Views last 7 days    0.0
Price_in_Euro        0.0
Country              0.0
City                 0.0
dtype: float64

In [43]:
# Checking the df
boat_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9241 entries, 0 to 9887
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Boat Type          9241 non-null   object 
 1   Manufacturer       9241 non-null   object 
 2   Type               9241 non-null   object 
 3   Year Built         9241 non-null   int64  
 4   Length             9241 non-null   float64
 5   Width              9241 non-null   float64
 6   Material           9241 non-null   object 
 7   Views last 7 days  9241 non-null   int64  
 8   Price_in_Euro      9241 non-null   int64  
 9   Country            9241 non-null   object 
 10  City               9241 non-null   object 
dtypes: float64(2), int64(3), object(6)
memory usage: 866.3+ KB


At this point the dataset is ready to be used

In [44]:
# Exporting the clean df
boat_data.to_csv(os.path.join(path, '02 Data','Prepared Data', 'boat_clean.csv'))
